Add generic support for bourne- and c-based shells

njalerikson opened the pull request:
With these changes conda's activate/deactivate now support bash/zsh/tcsh/csh.

The scripts assume that `activate*`, `deactivate*`, and `conda*` can be found on the user's `$PATH` as opposed to detecting the `$_SCRIPT_LOCATION` as was done previously. This change is necessary to support tcsh/csh where the script location cannot be **reliably** detected when sourced.

bash/zsh:

<pre>
source activate [ENVNAME]
source deactivate
</pre>


tcsh/csh (since the `which` command doesn't search the user `$PATH` in tcsh/csh):

<pre>
source \`which activate\` [ENVNAME]
source \`which deactivate\`
</pre>

## 

The old `activate` script has effectively been renamed as `activate.sh`. Likewise with the old `deactivate` script. They contain the same content but has been slightly edited to match the look and feel of the `*.csh` scripts that have been introduced.
## 

It may also make sense to encourage users to create aliases when they install Conda.

bash/zsh:

<pre>
alias activate="source activate"
alias deactivate="source deactivate"
</pre>


tcsh/csh:

<pre>
alias activate    "source \`which \activate\`"
alias deactivate  "source \`which \deactivate\`"
</pre>

msarahan commented:
This is a good idea, and a good approach for generally supporting more shells.  Please make sure you also add csh/tcsh to the test suite on relevant platforms.  See https://github.com/conda/conda/blob/master/utils/travis-run-script.sh#L12 for how to specify the shells to test.  You may need to also install the shells on Travis using apt-get?

The addition of aliases is something that the installer script should offer to do, IMHO, not just tell the user how to do it.

kalefranz commented:
Just noting that this is an awesome PR. Trying to get the bandwidth to spend some time with it.

njalerikson commented:
I realized that Travis CI runs on Ubuntu where the default shell is dash (yay). So I'm working to add dash support as well.

One problem with dash is that it doesn't support parameter passing to sourced scripts. My workaround is to have users set appropriate environment variables before sourcing:

```
CONDA_ENVNAME="envname" . activate
CONDA_HELP=true . activate
```

njalerikson commented:
Ok so I have addressed all of the unix errors.

Travis CI occasionally fails on the Mac tests due to test_cli.py/test_search5. I briefly looked at that error and determined that it appears to be unrelated to the work I have done and also that the error appears to be very hit and miss (I'll add a print statement to the test and it works, remove the print statement and it breaks).

I am aware that the AppVeyor fails and that is a result of the updates I have made to test_activate.py. Considering how I am not a Windows developer I am hesitant to start digging into that area. I do suspect however that most of the changes necessary to the Windows code would be related to the print/error messages and not any real functionality (besides the addition of the `-v,--verbose` feature).

I introduced 3 variables pertaining to activate/deactivate. In the below examples I show how to use one of the variables (`CONDA_ENVNAME`). The others are `CONDA_HELP` and `CONDA_VERBOSE`, both of which are boolean (true/false) values. All of this is explained in the expanded help dialog for activate and deactivate (see `. activate -h` and `. deactivate -h`).

@msarahan mentioned adding aliases to `source activate` to users' .profile at the same time as adding conda bin to the PATH. As far as I understand that is not something that is implemented in conda but rather in the actual Anaconda installation. Implementing such a change would mean that changes should also be made to conda/install.py/print_activate to detect whether or not an alias exists in the current shell and what shell is running. Checking for such things would be important in order to properly inform users of how to activate/deactivate given their setup:

<table align="center">
<tr align="center">
<th>Shell</th>
<th>w/ aliases</th>
<th>w/o aliases</th>
</tr>
<tr align="center">
<td><pre style="padding:0;"><code>csh/tcsh</code></pre></td>
<td><pre style="padding:0;"><code>activate [ENVNAME]</code></pre>
<p>or</p>
<pre style="padding:0;"><code>set CONDA_ENVNAME="ENVNAME"
activate</code></pre>
<p>then</p>
<pre style="padding:0;"><code>deactivate</code></pre></td>
<td><pre style="padding:0;"><code>source "`which activate`" [ENVNAME]</code></pre>
<p>or</p>
<pre style="padding:0;"><code>set CONDA_ENVNAME="ENVNAME"
source "`which activate`"</code></pre>
<p>then</p>
<pre style="padding:0;"><code>source "`which deactivate`"</code></pre></td>
</tr>
<tr align="center">
<td><pre style="padding:0;"><code>bash/zsh
(and sh on Mac OS)</code></pre></td>
<td><pre style="padding:0;"><code>activate [ENVNAME]</code></pre>
<p>or</p>
<pre style="padding:0;"><code>CONDA_ENVNAME="ENVNAME"
activate</code></pre>
<p>then</p>
<pre style="padding:0;"><code>deactivate</code></pre></td>
<td><pre style="padding:0;"><code>. activate [ENVNAME]</code></pre>
<p>or</p>
<pre style="padding:0;"><code>CONDA_ENVNAME="ENVNAME"
. activate</code></pre>
<p>then</p>
<pre style="padding:0;"><code>. deactivate</code></pre></td>
</tr>
<tr align="center">
<td><pre style="padding:0;"><code>dash
(and sh on Ubuntu)</code></pre></td>
<td><pre style="padding:0;"><code>CONDA_ENVNAME="ENVNAME"
activate</code></pre>
<p>then</p>
<pre style="padding:0;"><code>deactivate</code></pre></td>
<td><pre style="padding:0;"><code>CONDA_ENVNAME="ENVNAME"
. activate</code></pre>
<p>then</p>
<pre style="padding:0;"><code>. deactivate</code></pre></td>
</tr>
</table>

kalefranz commented:
@njalerikson sent you an email, but just something quick... For the bourne shells, should we just test against `posh`?

http://unix.stackexchange.com/a/48787/92065

njalerikson commented:
Given how I'm not even sure how to conveniently get `posh` onto my Mac I'm going to assume that it's not worth testing for on the Mac.

I did find it via `apt` on Linux.

njalerikson commented:
Hm an interesting gotcha, Travis CI doesn't support `posh` at this time (there's a whitelisting process). Maybe that means we don't want to support/test for posh?

njalerikson commented:
travis-ci/apt-package-whitelist#3278

kalefranz commented:
The master branch is now 4.3.   Let me know when you think this is ready @njalerikson, and we'll get it merged.

njalerikson commented:
Just an update, I'm working on an idea to also merge `fish` support into the logic of the c- and bourne- shell support. This will result in no longer requiring `fish` users to first source `conda.fish` and then use the activate/deactivate functions and instead activate/deactivate will look and work as it does on the other shells supported.

I'll also take a stab at the Windows support.

kalefranz commented:
> This will result in no longer requiring fish users to first source conda.fish

I actually really like how some of the fish stuff works, so don't go too far there just yet.

We pushed 4.2 to defaults today.  Getting this merged is a fairly high priority now for 4.3.  I've now got some bandwidth to spend time with it.

njalerikson commented:
I agree. I realized that `fish` actually has no way to be compatible with the design of the code that supports the other unix shells. A consequence of `fish` being non-POSIX.

As a result the `fish` code is unchanged although it may be important to add unittests for fish.

I have instead started to work on the Windows scripts in order update their abilities to be inline with the functionality of the unix shells. Effectively just trying  to fix up the windows codes to pass the tests at this point (yay for test driven development).

I should get some time in the next few days to get something put together.

dafh commented:
Hi everyone, thanks for your hard work for C-shell people like me. I'm kind of new with conda (I'm using miniconda by the way) and github, but can you tell me which commit should I follow to use source activate on tcsh? (I'm kind of confused here)

njalerikson commented:
It is still from my private fork of Conda (https://github.com/njalerikson/conda).

The most recent commit is focused on Windows support, the unix support has been pretty aggressively tested to date so you should be safe working with that.

dafh commented:
Thanks for your answer: 

I cloned your master branch, and I did: 'python setup.py build' but I got this:

WRITING build/lib/conda/.version with version None
Traceback (most recent call last):
  File "setup.py", line 98, in <module>
    zip_safe=False,
  File "/home/dafh/miniconda2/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/home/dafh/miniconda2/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/home/dafh/miniconda2/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/home/dafh/miniconda2/lib/python2.7/distutils/command/build.py", line 127, in run
    self.run_command(cmd_name)
  File "/home/dafh/miniconda2/lib/python2.7/distutils/cmd.py", line 326, in run_command
    self.distribution.run_command(command)
  File "/home/dafh/miniconda2/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/home/dafh/conda-master/conda/_vendor/auxlib/packaging.py", line 206, in run
    write_version_file(target_dir, self.distribution.metadata.version)
  File "/home/dafh/conda-master/conda/_vendor/auxlib/packaging.py", line 198, in write_version_file
    f.write(version)
TypeError: expected a string or other character buffer object

So I'm assuming I'm doing something wrong (or missing about .version). Is there a README with instructions-for-dummies (me in this case)?

njalerikson commented:
I personally install any Python package like this using `pip` which is the modern standard for Python:

```
cd <CONDA_DIR>
pip install .
```

Second (as a general Python pointer), you should not be using `python setup.py build` unless you are looking to deploy something you have coded up. You will almost always want to be using `install`:

```
python setup.py install
```

Details on `build`: https://docs.python.org/3.5/distutils/builtdist.html
Details on `python setup.py install`: https://docs.python.org/3.5/install/
Details on `pip install .`: https://docs.python.org/3.5/installing/index.html#installing-index

dafh commented:
Yes, the thing is 'pip install .' tried to build (wheel?) and it failed. Then I did python setup.py build and the same error (the one I wrote in the last comment). Now I did ' python setup.py install', it goes (apparently) without error, but when I do: 'activate' this appears:

awk: fatal: can't open source file `whichshell.awk: Command not found.' for reading (No such file or directory)
whichshell_args.bash: Command not found.
awk: fatal: can't open source file `whichshell.awk: Command not found.' for reading (No such file or directory)
whichshell_args.bash: Command not found.
awk: fatal: can't open source file `whichshell.awk: Command not found.' for reading (No such file or directory)
whichshell_args.bash: Command not found.

I know I can find those files in shell/ on your master-branch, but it is clear for me I messed up with some stuff. Thanks for the 'build' advice!.

dafh commented:
You are absolutely right! Now it works. Thanks a lot.

dafh commented:
Hi njalerikson, a brief question: could it be that the env variable $CONDA_PREFIX is missed somehow in tcsh? I have 'iraf27' environment, so if I use '. activate iraf27' with bash, that env variable is setted up correctly. Then, if I use tcsh and I enter to iraf27 (source \`which \activate\` iraf27 ), it is not setted up (and I cannot use IRAF cause iraf.sh [like a setup code] uses it).

njalerikson commented:
I am not aware of the issue you have. I do expect the $CONDA_PREFIX to be removed upon deactivate.

```
>
> tcsh
>
> source "`which activate`" py2.7
(py2.7) >
(py2.7) > echo $CONDA_PREFIX
~/anaconda3/envs/py2.7
(py2.7) >
(py2.7) > source "`which deactivate`" 
> 
> echo $CONDA_PREFIX 
CONDA_PREFIX: Undefined variable.
> 
```

dafh commented:
Yes! I was confused because for 'some reason' IRAF doesn't work in iraf27 env with tcsh (but do work in bash), but the env variables are setted up in tcsh version so that's not the issue. Sorry!

njalerikson commented:
@kalefranz 

I've been looking to finish up the Windows support. I think it makes sense to revisit previous decisions to support Windows activate/deactivate without symlinking (currently `win_conda_bat_redirect` is used instead of symlinks on Windows). Python supports symlinking for Windows 6 (Vista) and newer (https://docs.python.org/3.5/library/os.html#os.symlink). Not sure how much legacy support Conda is expected to provide but choosing to drop anything older than Windows 6 (Vista) support will drastically simplify activate/deactivate on Windows.

This will also open up opportunities to support other important shells (e.g. PowerShell).

Even AppVeyor uses Windows Server 2012 (which corresponds to Windows 8).

This should especially be revistited since according to Anaconda's FAQ only version 2.2 supports Windows XP. So all "modern" Windows since XP should support symlinking.

kalefranz commented:
I need to catch up on this thread. But yes we only support back to XP. 

I *really* want to do away with all of the symlinking conda on activate.  

> On Oct 6, 2016, at 10:02 AM, Ken Odegard <notifications@github.com> wrote:
> 
> @kalefranz
> 
> I've been looking to finish up the Windows support. I think it makes sense to revisit previous decisions to support Windows activate/deactivate without symlinking (currently win_conda_bat_redirect is used instead of symlinks on Windows). Python supports symlinking for Windows 6 (Vista) and newer (https://docs.python.org/3.5/library/os.html#os.symlink). Not sure how much legacy support Conda is expected to provide but choosing to drop anything older than Windows 6 (Vista) support will drastically simplify activate/deactivate on Windows.
> 
> This will also open up opportunities to support other important shells (e.g. PowerShell).
> 
> Even AppVeyor uses Windows Server 2012 (which corresponds to Windows 8).
> 
> This should especially be revistited since according to Anaconda's FAQ only version 2.2 supports Windows XP. So all "modern" Windows since XP should support symlinking.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mingwandroid commented:
@njalerikson Using symlinks on Windows is not something we can do, from the Python documentation:

`
On Windows, the SeCreateSymbolicLinkPrivilege is required in order to successfully create symlinks. This privilege is not typically granted to regular users but is available to accounts which can escalate privileges to the administrator level. Either obtaining the privilege or running your application as an administrator are ways to successfully create symlinks.
OSError is raised when the function is called by an unprivileged user.
`

It is very complicated to get that privilege on a computer on which you are an Administrator (due to Windows security token filtering).

It is (in my experience) nigh-on impossible to convince your IT department to give you that privilege on an enterprise computer.

kalefranz commented:
Wow there's something up with that merge diff.  It's definitely not tracking conda/conda:master.  Trying to figure out what's going on...

njalerikson commented:
yeah looks like I broke something somewhere, you should probably kill those builds

kalefranz commented:
Yeah I think it was just that last merge commit. 3777d43

Probably if you do

    git reset --hard 3910739
    git cherry-pick 94d4bc2

Might be the easiest way.  Then we'll figure out the merge conflicts if there are any with master.

njalerikson commented:
Yeah I just arrived to the same conclusion, I'm guessing my GitHub Desktop wasn't completely in sync with the most current master so that big merge just messed it all up

bkreider commented:
Will this also  work for Korn shell?

kalefranz commented:
Yeah I think @njalerikson explicitly added ksh as a test vector.

kalefranz commented:
Hey @njalerikson I wouldn't revert.  I'd back up, cherry-pick, and then force push.  The diff still doesn't look right.

kalefranz commented:
@njalerikson this code is immaculate.  Well done.

Reading through the code, I don't see any reason not to merge this at this point.  Probably best for you to fix those three merge conflicts instead of me.  I promise not to merge anything else to master that could break this PR until we get it merged :-D

Only one comment reading through the code so far...  I'd rather not put all of those executables in `bin`.  While the main `activate` and `deactivate` entry points stay in `bin`, I think the correct place for the helper/shell-specific scripts is `libexec`.  The [wiki FHS page](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) doesn't mention `libexec`, so I guess it may be somewhat non-standard.  Maybe it's just a [red hat](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/3/html/Reference_Guide/s1-filesystem-fhs.html) convention.

> The `libexec/` directory contains small helper programs called by other programs

I'd also be fine just putting them in `lib`.  Or another suggestion if you have it.  I just don't want to clutter up people's `bin` directory more than we already have :)

kalefranz commented on shell/envvar_cleanup.bash:
```diff
@@ -0,0 +1,327 @@
+#!/usr/bin/env bash
```
This comment shouldn't block merging this PR, but...  I'm pausing and thinking a little bit here.  This effectively makes bash a hard-dependency of conda.  (Not that it wasn't before, but this PR is all about cleaning things up!)

It's not jumping out at me right now, but I'm guessing you use some bash-ness somewhere that's not strictly posix?

I doubt anybody will care, but this could come up as an issue.  And if it does, it seems to me there are several possibilities on how to proceed.  The most obvious one to me would be to write a separate entry_point or `conda ..XXXXX` helper command.  Obviously that returns *quickly* and doesn't load a ton of conda code imports.  Then we could also unify the duplicated logic that's in the .bat version.

kalefranz replied:
BTW, I actually think I'm ok with `awk` being a hard dependency for unix systems.  Just as a reference I checked an instance I have running busybox, and awk is part of the busybox binary.  I guess I'd generally consider busybox a good data point for what might be possible to make hard dependencies.

kalefranz commented:
Just checked out the PR to start running tests and playing with the code after reading through it.  Sort of having trouble 😬 

```
kfranz@0283:~/continuum/conda *pr-3175 ❯ ./shell/test_whichshell
TEST_WHICHSHELL

WHICHSHELL_PS.BASH
./shell/test_whichshell: line 24: whichshell_ps.bash: command not found
IS (t)csh
./shell/test_whichshell: line 25: whichshell_ps.bash: command not found
csh

WHICHSHELL_ARGS.BASH: ERROR-CLAUSE
./shell/test_whichshell: line 32: whichshell_ps.bash: command not found
./shell/test_whichshell: line 32: whichshell_args.bash: command not found

WHICHSHELL_ARGS.BASH: BOURNE-SHELL-CLAUSE
./shell/test_whichshell: line 37: whichshell_ps.bash: command not found
./shell/test_whichshell: line 37: whichshell_args.bash: command not found

WHICHSHELL_ARGS.BASH: C-SHELL-CLAUSE
./shell/test_whichshell: line 42: whichshell_ps.bash: command not found
./shell/test_whichshell: line 42: whichshell_args.bash: command not found

WHICHSHELL.AWK: ERROR-CLAUSE
./shell/test_whichshell: line 50: whichshell_ps.bash: command not found
./shell/test_whichshell: line 50: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)

WHICHSHELL.AWK: BOURNE-SHELL-CLAUSE
./shell/test_whichshell: line 56: whichshell_ps.bash: command not found
./shell/test_whichshell: line 56: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)

WHICHSHELL.AWK: C-SHELL-CLAUSE
./shell/test_whichshell: line 62: whichshell_ps.bash: command not found
./shell/test_whichshell: line 62: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)



kfranz@0283:~/continuum/conda *pr-3175 ❯ cd shell/
kfranz@0283:~/continuum/conda/shell *pr-3175 ❯ ./test_whichshell
TEST_WHICHSHELL

WHICHSHELL_PS.BASH
./test_whichshell: line 24: whichshell_ps.bash: command not found
IS (t)csh
./test_whichshell: line 25: whichshell_ps.bash: command not found
csh

WHICHSHELL_ARGS.BASH: ERROR-CLAUSE
./test_whichshell: line 32: whichshell_ps.bash: command not found
./test_whichshell: line 32: whichshell_args.bash: command not found

WHICHSHELL_ARGS.BASH: BOURNE-SHELL-CLAUSE
./test_whichshell: line 37: whichshell_ps.bash: command not found
./test_whichshell: line 37: whichshell_args.bash: command not found

WHICHSHELL_ARGS.BASH: C-SHELL-CLAUSE
./test_whichshell: line 42: whichshell_ps.bash: command not found
./test_whichshell: line 42: whichshell_args.bash: command not found

WHICHSHELL.AWK: ERROR-CLAUSE
./test_whichshell: line 50: whichshell_ps.bash: command not found
./test_whichshell: line 50: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)

WHICHSHELL.AWK: BOURNE-SHELL-CLAUSE
./test_whichshell: line 56: whichshell_ps.bash: command not found
./test_whichshell: line 56: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)

WHICHSHELL.AWK: C-SHELL-CLAUSE
./test_whichshell: line 62: whichshell_ps.bash: command not found
./test_whichshell: line 62: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)



kfranz@0283:~/continuum/conda *pr-3175 ❯ ls -al /conda3/bin/
total 4496
drwxr-xr-x  44 kfranz  wheel   1.5K Nov  8 09:02 ./
drwxr-xr-x  12 kfranz  wheel   408B Jul 29 02:26 ../
-rwxrwxr-x   2 kfranz  wheel    13K May 31 16:36 .openssl-libcrypto-fix*
-rw-rw-r--   2 kfranz  wheel   225B May 31 16:36 .openssl-post-link.sh
-rwxr-xr-x   1 kfranz  wheel    95B Jul 28 14:16 2to3*
-rwxr-xr-x   1 kfranz  wheel   3.4K Nov  8 09:02 activate*
-rwxr-xr-x   1 kfranz  wheel    14K Nov  8 09:02 activate.csh*
-rwxr-xr-x   1 kfranz  wheel    13K Nov  8 09:02 activate.sh*
-rwxrwxr-x   1 kfranz  wheel   5.0K Jul 28 14:16 c_rehash*
-rwxr-xr-x   1 kfranz  wheel   339B Nov  8 09:02 conda*
-rwxrwxr-x   1 kfranz  wheel    90B Jul 29 02:26 conda-build*
-rwxrwxr-x   1 kfranz  wheel    92B Jul 29 02:26 conda-convert*
-rwxrwxr-x   1 kfranz  wheel    92B Jul 29 02:26 conda-develop*
-rwxrwxr-x   1 kfranz  wheel    90B Jul 29 02:26 conda-index*
-rwxrwxr-x   1 kfranz  wheel    92B Jul 29 02:26 conda-inspect*
-rwxrwxr-x   1 kfranz  wheel    96B Jul 29 02:26 conda-metapackage*
-rwxrwxr-x   1 kfranz  wheel    93B Jul 29 02:26 conda-pipbuild*
-rwxrwxr-x   1 kfranz  wheel    91B Jul 29 02:26 conda-render*
-rwxrwxr-x   1 kfranz  wheel    89B Jul 29 02:26 conda-sign*
-rwxrwxr-x   1 kfranz  wheel    93B Jul 29 02:26 conda-skeleton*
-rwxr-xr-x   1 kfranz  wheel   3.4K Nov  8 09:02 deactivate*
-rwxr-xr-x   1 kfranz  wheel    11K Nov  8 09:02 deactivate.csh*
-rwxr-xr-x   1 kfranz  wheel    10K Nov  8 09:02 deactivate.sh*
-rwxr-xr-x   1 kfranz  wheel   138B Jul 28 14:15 easy_install*
lrwxr-xr-x   1 kfranz  wheel    12B Jul 28 14:15 easy_install-2.7@ -> easy_install
-rwxr-xr-x   1 kfranz  wheel    13K Nov  8 09:02 envvar_cleanup.bash*
-rwxr-xr-x   1 kfranz  wheel    93B Jul 28 14:16 idle*
-rwxrwxr-x   2 kfranz  wheel   547K May 31 16:36 openssl*
-rwxr-xr-x   1 kfranz  wheel   110B Jul 28 14:15 pip*
-rwxr-xr-x   1 kfranz  wheel    78B Jul 28 14:16 pydoc*
lrwxr-xr-x   1 kfranz  wheel     9B Jul 28 14:16 python@ -> python2.7
lrwxr-xr-x   1 kfranz  wheel    16B Jul 28 14:16 python-config@ -> python2.7-config
lrwxr-xr-x   1 kfranz  wheel     9B Jul 28 14:16 python2@ -> python2.7
lrwxr-xr-x   1 kfranz  wheel    16B Jul 28 14:16 python2-config@ -> python2.7-config
-rwxr-xr-x   1 kfranz  wheel   8.8K Jul  2 17:44 python2.7*
-rwxr-xr-x   1 kfranz  wheel   1.6K Jul 28 14:16 python2.7-config*
-rwxr-xr-x   1 kfranz  wheel    18K Jul 28 14:16 smtpd.py*
-rwxr-xr-x   2 kfranz  wheel   1.4M May 25 17:33 sqlite3*
-rwxr-xr-x   2 kfranz  wheel    12K Mar 16  2015 tclsh8.5*
-rwxr-xr-x   1 kfranz  wheel   117B Jul 28 14:15 wheel*
-rwxr-xr-x   1 kfranz  wheel   8.4K Nov  8 09:02 whichshell.awk*
-rwxr-xr-x   1 kfranz  wheel   3.5K Nov  8 09:02 whichshell_args.bash*
-rwxr-xr-x   1 kfranz  wheel   7.2K Nov  8 09:02 whichshell_ps.bash*
-rwxr-xr-x   2 kfranz  wheel    25K Mar 16  2015 wish8.5*


kfranz@0283:~/continuum/conda/shell *pr-3175 ❯ source /conda3/bin/activate /conda3
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)



kfranz@0283:~/continuum/conda/shell *pr-3175 ❯ source /conda3/bin/activate root
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)



kfranz@0283:~/continuum/conda/shell *pr-3175 ❯ source /conda3/bin/activate -h
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)
-bash: whichshell_ps.bash: command not found
-bash: whichshell_args.bash: command not found
awk: fatal: can't open source file `(null)' for reading (reason unknown)
```

kalefranz commented:
Nevermind.  Solution was just `export PATH="/conda3/bin:$PATH"`

njalerikson commented:
I have fixed a lot of issues in my local git, granted most of that is related to the windows issues.

As far as I can tell `envvar_cleanup.bash` is POSIX compatible. The only reason I force it to `bash` is because the consistency is comforting (I've learned to not trust `sh`).

kalefranz commented:
I can't blame you there on not trusting `sh`.  Let's see how it goes I guess.

kalefranz commented:
Another note... with `test_activate.py`, I'd like to get it to a point where conda doesn't have to be installed for the tests to pass.  That's actually the reason the tests in Travis-CI are split right now.

I use homebrew python a lot locally.  So with homebrew python and pip-installed dependencies--and no conda, activate, or deactivate on `PATH`--it would be ideal if `python -m pytest tests/test_activate.py --shell=bash` passed.  I would probably start by adding `setUp` and `tearDown` methods to the test harness that adds and removes `shell` to/from `PATH`.

kalefranz commented:
@njalerikson I'd like to get this merged sooner rather than later so I can stop breaking your PR with merge conflicts.  We can always iterate, refactor, and improve with future PRs.

kalefranz commented:
Added a couple of tags to indicate this will be merged before the 4.3.0 release goes out.

kalefranz commented:
@njalerikson From your perspective, are you ok merging this?  And then making future PRs from there?  I'll take it over and fix the merge conflicts if you want.  I'll also make that switch from `bin` to `libexec` that I was talking about for certain files.

njalerikson commented:
@kalefranz So the AppVeyor tests (same code) passed no problem when I ran it earlier (not part of this PR). It is currently failing with "there are newer queued builds for this pull request, failing early".

Is this somehow related to how I had to force push?

kalefranz commented:
Could have been.  I just hit the rebuild button on Appveyor.

njalerikson commented:
@kalefranz I think I have addressed all of my concerns for merging.

There seems to be a backlog on Travis' macOS builds so if you have the opportunity you may want to kill the old jobs from this PR that are still queued.

I see that one test from `test_create` seems to be consistently failing, I have not dug into the details.

kalefranz commented:
Just FYI, there's a failing test `test_install_tarball_from_local_channel` right now pretty much everywhere.  Nothing to do with this PR.  The failure got triggered with an update to the `requests` library in the default channel.  Took a day and a half to track down :-/

It'll be fixed in master soon.

kalefranz commented:
> I see that one test from test_create seems to be consistently failing, I have not dug into the details.

Oh, yeah.  What I said above ^^^

Really glad you didn't waste time digging into it :)

kalefranz commented:
If everything passes here except that one test, I'm hitting the merge button!

njalerikson commented:
I suppose the only danger is that since the failing test is a test that occurs in the not installed test group none of the travis tests are actually running the activate tests that this PR cares about...

kalefranz commented:
Yes, I just thought about that

The fix for the test failure is (hopefully) in this PR: https://github.com/conda/conda/pull/3886
The fix obviously has nothing to do with the purpose of the PR.  It was just as good of place as any to iterate on it since the original PR was so simple.

When that's merged, let's merge master into this branch, make sure the activate tests pass, and then we'll be good to go.

kalefranz commented:
Ok go ahead and merge master.  Should be good now.  I'll try to clear out the Travis-CI queue to get this running.

kalefranz commented:
The Travis-CI OSX backlog is huge right now.  https://www.traviscistatus.com/

I ran `py.test -m "installed" tests --shell=bash` locally and everything passed.

So, 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

kalefranz commented:
Do you ever come to Austin?  I owe you a drink.  Or three.  Or more.

njalerikson commented:
I'll hold you to that. If not before I may have to make AnacondaCON work.

mingwandroid commented:
Sorry this is very late, but while testing this I ran into an issue with:

> The scripts assume that activate*, deactivate*, and conda* can be found on the user's $PATH as opposed to detecting the $_SCRIPT_LOCATION as was done previously. This change is necessary to support tcsh/csh where the script location cannot be reliably detected when sourced.

Isn't the largest part of the point of activate to set the user's path? Could we make it so that people who use the good shells (i.e. `bash` or `zsh`) do not have to jump through hoops so that we can support less good shells?

mingwandroid commented:
.. also, forcing the user to add this to their PATH results in that PATH always being present:

Here I created a new env, `test_env` and ran `source activate test_env`, but I still also have the root env on my PATH (`/Users/ray/mc-x64-2.7/bin`). My final PATH is:

`/Users/ray/mc-x64-2.7/envs/test_env/bin:/Users/ray/mc-x64-2.7/bin:/usr/local/texlive/2016/bin/x86_64-darwin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/opt/X11/bin:/usr/X11/bin:/Library/TeX/texbin`

This will cause all sorts of behavior changes (most worrying for me, when `conda-build` runs its tests it will find dependencies from the root env and return a false positive).

I'm not *necessarily* -1 on having root always on PATH in general (I am undecided) but I don't think it should be done as a side-effect of supporting less good shells and I definitely think it must be optional (for example, during testing).

wulmer commented:
Missed parts of the "root always on PATH" discussion, but for me (and several thousands other users in my company on Windows) it would be acceptable to have the root env on the PATH *until* the user activates a non-root environment. Windows users are not used to caring about PATH setup, and hardly anyone knows about the `where` command (similar to `which`). @njalerikson Do you see a possibility to keep the support for all those shells while making root disappear after an environment got activated?

njalerikson commented:
As discussed in other issues/PRs and mulled over on my own I believe the correct thing to do is not to have the root environment on the user's path. I instead believe that the correct thing to do is to introduce a `sbin` directory under the conda directory. The path to conda's `sbin` or similar must always be on the user's `PATH`. This `sbin` will only include the bare minimum conda tools that are necessary for any basic conda functions (this would predominantly be the `activate`/`deactivate` scripts).

@mingwandroid As a general good practice for OS/shell agnostic tools, I see little value in making different shells behave differently (whether for good vs bad OSes or good vs bad shells). This adds both a confusion on the user's side where different documentation is necessary for different people but also drastically complicates long-term support for code when code that is intended to mirror each other no longer does so. Further I do think the `sbin` idea will address most real concerns with forcing some conda directory onto the user's `PATH`.

@wulmer The main reason I was forced to leave root on the PATH is because certain shells are simply unable to detect script locations when sourced and hence I would have no way to detect the root path (unless we introduced some new environment variable like `CONDA_ROOT`). I feel like the `sbin` concept would address your concerns.

wulmer commented:
@njalerikson Yes, sounds like the `sbin` concept solves the problem with accidentally having multiple envs on the PATH. I don't have a problem with the root env *not* being on the user's PATH per default. It is easy to add it if required. But confusing the user with running applications from different envs isn't a good choice for new conda users. BTW: Thanks for this awesome PR!

mingwandroid commented:
It is an awesome PR and I am very appreciative of the hard work you've done.

This is of course up to @kalefranz and yourself, but do you think it is possible to get the `sbin` change in before 4.3? If not, I think the old behaviour needs to be restored. Keeping the root env on PATH will lead to tests passing that should not and therefore people will create broken recipes and broken packages.

msarahan commented:
Please also keep legacy users in mind. Any solution needs to not require
users to modify their existing PATH setting. Adding to new installs is
fine, but there needs to be a compatible fallback for old installs.

This is what went extremely wrong with the Anaconda 2.4 release last year.
We moved many DLLs into library/bin and added that folder to PATH in new
installs. However, old installs did not have it in PATH, and disaster
ensued.

On Fri, Nov 18, 2016, 16:15 Ray Donnelly <notifications@github.com> wrote:

> It is an awesome PR and I am very appreciative of the hard work you've
> done.
>
> This is of course up to @kalefranz <https://github.com/kalefranz> and
> yourself, but do you think it is possible to get the sbin change in
> before 4.3? If not, I think the old behaviour needs to be restored. Keeping
> the root env on PATH will lead to tests passing that should not and
> therefore people will create broken recipes and broken packages.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/conda/conda/pull/3175#issuecomment-261656330>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACV-RZ76QH0pUzYhzHGWpGOPNllNCJyks5q_iN9gaJpZM4JWodC>
> .
>

kalefranz commented:
How about, with `4.3`, we communicate a new way to install conda.  We'd do this loudly and obnoxiously to make sure people got the message.  We can also make it more programatic and extensible so that it hopefully wouldn't have to change for years to come.

I've found two examples in my shell init, and I'm sure there are many others we could cheat off of.

```
# Load RVM into a shell session *as a function*
[[ -s "$HOME/.rvm/scripts/rvm" ]] && source "$HOME/.rvm/scripts/rvm"
```

and

```
export RBENV_ROOT=/usr/local/var/rbenv
if which rbenv > /dev/null; then eval "$(rbenv init -)"; fi
```

Yes, this could potentially break people with an upgrade, as @msarahan has warned about.  We can also anticipate this though and handle it.  By handle it, I mean we can at a minimum give warning messages.  We could go so far as attempting to add the needed lines to the needed shell init scripts too with post-link scripts (of course in this case we'd still loudly communicate what we're doing so we minimize surprises).

@mingwandroid if we were to go down this route, how do we do this on windows?  Is there some place we can hook in for `cmd.exe` init?

njalerikson commented:
I have mulled over the option to jumpstart the `sbin` changes (would still require a major change to user's `PATH`) or whether to work on adding a small check method to conda that probes the user's `PATH` to see if their environment is properly configured with the conda location (and if not prompt and inform of the 4.3 requirement).

I feel that a check method is the better route to go since I do not see any way to allow conda to exist unrelated to the `PATH` especially given the ideal direction of conda.

Modifying the windows path is very "fun":
**manually**: https://www.java.com/en/download/help/path.xml
**coded**: http://stackoverflow.com/a/8358361/5239932

mingwandroid commented:
Are you actually contemplating changing the user's System Path? IMHO, software that modifies the state of the OS in such a fundamental way is obnoxious (other software will no longer work).

Path changes need to be as localized as possible. Please fix it so that the activate script can be launched from a full path as it was before. From my searching around, both `csh` and `tsch` can identify the full path of the script they are currently running.

njalerikson commented:
I am/was contemplating appending to the **user's** variable PATH. That is still something set outside of the cmd.exe session but is not a system variable change.

But very well, given my understanding the best I can do that is somewhat safe is to derive the path of sourced files from the process' ID (if an absolute/relative path is given use that, otherwise assume conda is on the `PATH`). I can justify this since the process details are already in use to detect the user's shell. Otherwise we would need to introduce a new environment variable (`CONDA_ROOT`) to tell the scripts where to find stuff.

It is not just csh/tcsh that is difficult to work with when sourcing. Remember we are currently officially supporting the following shells:
bash, dash, posh, zsh, ksh, csh, tcsh, cmd.exe

Conditionally supporting:
PowerShell, Cygwin, MinGW, MSYS

If you have enlightening ways to detect the path to sourced files in a cross platform and cross shell manner I would love to learn.

mingwandroid commented:
> I am/was contemplating appending to the user's variable PATH. That is still something set outside of the cmd.exe session but is not a system variable change

What I meant here is that for that user, such a change would be system-wide (to my mind, if it's stored in the Windows Registry, then it is a system issue) and it *will* break other software. IMHO this is not something we can even begin to contemplate. On Linux and OS X, software uses RPATHs to make PATH changes less of an issue* but Windows does not have any equivalent concept unfortunately.

> the best I can do that is somewhat safe is to derive the path of sourced files from the process' ID (if an absolute/relative path is given use that, otherwise assume conda is on the PATH).

bash, dash, posh, zsh, ksh, csh, tcsh: In general, on any Unix shell you have such fallbacks as those listed here: http://stackoverflow.com/questions/2563300/how-can-i-find-the-location-of-the-tcsh-shell-script-im-executing
For a batch file: %~dp0
Cygwin, MSYS2: Depends on the Unix shell, is no different from the same shell on Unix modulus perhaps calling `$(cygpath -wm)` or `$(cygpath -w)`.
PowerShell: I don't know, I never use this.

> If you have enlightening ways to detect the path to sourced files in a cross platform and cross shell manner I would love to learn.

OK, another alternative I just cooked up that works on both macOS and Linux:

```
#!/usr/bin/env bash

ME=$(lsof -p $$ | grep activate.sh | grep -o '[^ ]*$')
echo ME is $ME
```

> Otherwise we would need to introduce a new environment variable (CONDA_ROOT) to tell the scripts where to find stuff

We already have this, it is called `${CONDA_PREFIX}`

*Even on Linux or macOS, I would not recommend a user of Anaconda/Minconda to add it to their PATH in a startup script because you are then at the mercy of how well behaved other software is.

kalefranz commented:
Was about to add something @mingwandroid just did...  Tested on `dash`, and `dirname "$0"` should always give the path of the script's directory.

njalerikson commented:
**Sourcing** is the name of the game.

Sourcing and executing scripts are very different animals.

This goes into extensive discussion concerning this: http://stackoverflow.com/questions/29832037/how-to-get-script-directory-in-posix-sh

While there is no POSIX defined standard for getting the path of a sourced script most of the shells offer some solution:
* `$BASH_SOURCE` (only for bash)
* `${.sh.file}` (only for ksh)
* `$_` (mixed results, in cases where this works it will only do so if it is run as the very first command in the sourced script)

These checks would effectively need to occur in the top level activate/deactivate scripts since the top level scripts currently call several other vital scripts. Viewing this file and you will see that no variable assignments occur there. This is because the bourne and c families do not set variables in compatible ways. This consequently means that many of the function calls are repeatedly made and while the top level activate/deactivate do perform an if-elseif-else logic it uses nothing that looks like an if-clause.

I firmly believe that `$CONDA_ROOT` is the way to go in order to quickly resolve this. The `$CONDA_PREFIX` variable is related to the activate environment. **It is set by the activate script and is likewise unset by the deactivate script.** I am proposing a new environment variable that will persist beyond activate/deactivate and always refer to the conda install location. This will also allow for the option to remove the root environment upon activate of a non-root environment (so killing two birds with one stone).

I encourage you to look over the top level activate and deactivate scripts. You should not have to look much past that file before understanding the inherent complexity in supporting these two shell families.

kalefranz commented:
> **Sourcing** is the name of the game.

:facepalm:

kalefranz commented:
@njalerikson Does `eval` work on all bourne shells? And is there an equivalent in all c shells?

njalerikson commented:
Yes `eval` is cross-shell (even `fish` has `eval`!).

kalefranz commented:
I don't mind `CONDA_ROOT`.  It's sorta been [used in the past](https://github.com/conda/conda/blob/4.0.x/conda/config.py#L129-L133) with a comment-warning that you really shouldn't use it.  

I'm using it officially now: https://github.com/conda/conda/blob/4.2.x/conda/__init__.py#L30-L31

Setting it elsewhere shouldn't affect anything though, as long as you set it "correctly" (i.e. to an *actual* conda root).

The more time I spend on this (I've been hacking around for several hours this evening), the more I'm liking something along the lines of 

    export CONDA_ROOT=/home/me/anaconda
    # or:   setenv CONDA_ROOT /home/me/anaconda
    eval "$(conda init -)"

At that point, there are SOOO many things we can do to make our lives easier and our users lives better.  Things like `conda activate` [#341, #2080, #2337, #3111].  We don't have to get this ambitious yet, but *someday* we're going to have to communicate a change in the install procedure.  This is what it will be (or something similar) when we communicate it.  From there, users should never have to change again, because we have control of the code from there.

kalefranz commented:
@njalerikson One thing we absolutely need to fix, pointed out by @mingwandroid 

> Please fix it so that the activate script can be launched from a full path

This is really important.  *So* much code--not just the conda and conda-build code--uses full paths for reliability.  A lot of devops code for example.  Heck, I typically full-path both the `conda` and `activate` executables because I have so many different versions installed locally.  We'll need to figure out a solution here.

kalefranz commented:
... another possibility...  Use conda's own magic of rewriting prefixes in text files!  Seems like such an obvious solution.

http://conda.pydata.org/docs/building/meta-yaml.html#text-files-with-prefix-files

We can list whatever text files we want to hard-code full-path in `build:has_prefix_files`.

Just another possibility.

njalerikson commented:
I have several ideas ranging from the aforementioned `CODNA_ROOT` idea to completely rewriting the `whichshell*` in Python. None of these options appear to be obviously correct and I feel like I'll end up walking in circles, at least with rewriting `whichshell*`.

I do like the idea of rewriting prefixes. Here's a list of all the scripts/executables that are currently assumed can be found on the `PATH`.

```
whichshell* 
activate*
deactivate*
envvar_cleanup*
conda
```

None of these files would have competing or conflicting with other versions within a conda tree.

kalefranz commented:
I'm hacking around with various ideas here too. Want to try to schedule a chat for tomorrow to bounce things around?

> On Nov 22, 2016, at 9:23 AM, Ken Odegard <notifications@github.com> wrote:
> 
> I have several ideas ranging from the aforementioned CODNA_ROOT idea to completely rewriting the whichshell* in Python. None of these options appear to be obviously correct and I feel like I'll end up walking in circles, at least with rewriting whichshell*.
> 
> I do like the idea of rewriting prefixes. Here's a list of all the scripts/executables that are currently assumed can be found on the PATH.
> 
> whichshell* 
> activate*
> deactivate*
> envvar_cleanup*
> conda
> None of these files would have competing or conflicting with other versions within a conda tree.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mingwandroid commented:
The gold standard for deactivation is that it should cause all modifications to the environment to be undone (because that's the only reasonable definition of deactivation since conda is uni-env; even if conda were multi-env, deactivating all the envs should undo all the modifications). I'll try to find some time to work on sorting out `BASH_SOURCE[0]` for when sourcing in poor shells (I make no apologies for calling them that, this is a basic feature).

njalerikson commented:
@mingwandroid `deactivate` almost completely reverses the original activation. There are a few caveats that may cause slight reordering of the `PATH` (but only the conda related paths are reordered, nothing else from the user's `PATH` is touched). Now, unlike previous versions of conda, if the user modifies their `PATH` while in an environment and then calls `deactivate` only the conda environment on the `PATH` is removed/updated, the other non-conda modified parts remain as the user changed.

Of course they are lesser shells, I really do not expect anyone to work in `dash` (as an aside there are special places for those who use spaces in pathnames and also work in `csh`/`tcsh`). What I do see is that each time I was able to incorporate a new shell I found that I was able to create a more robust solution. Besides it's cool to call Conda OS **and** shell-agnostic :)

njalerikson commented:
@kalefranz sent you an email yesterday but I seem to have a better shot of contacting you via a post here...

StevenCHowell commented:
@njalerikson I just came across an incompatibility with a colleague who uses tcsh.  After reviewing this issue, it seems the issue be resolved, but I can reproduce the failing behavior even with the most recent anaconda version, 4.3.1-np111py36-0.

Here is the behavior I see
```
odin:~/data/sync/nrc% source activate jlab
activate: No such file or directory.
```
`activate` seems to not be found
```
odin:~/data/sync/nrc% which activate
/home/schowell/data/myPrograms/anaconda3/bin/activate
```
even though `which` can find it.  When I use the full path for activate I get the following
```
odin:~/data/sync/nrc% source `which activate` jlab
BASH_VERSION: Undefined variable.
```
Is this expected?

njalerikson commented:
The changes outlined in this PR do not exist in the any of the official conda releases at this time.

You should find that running `ls /home/schowell/data/myPrograms/anaconda3/bin/` will only have one activate and one deactivate script and they will not look [this](https://github.com/conda/conda/blob/master/shell/activate) and [this](https://github.com/conda/conda/blob/master/shell/deactivate).
